sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestStream.java

```diff
@@ -0,0 +1,326 @@
+ */
+
+package org.apache.beam.sdk.testing;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.DurationCoder;
+import org.apache.beam.sdk.coders.InstantCoder;
+import org.apache.beam.sdk.coders.IterableCoder;
+import org.apache.beam.sdk.coders.StandardCoder;
+import org.apache.beam.sdk.runners.PipelineRunner;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.util.PropertyNames;
+import org.apache.beam.sdk.util.VarInt;
+import org.apache.beam.sdk.util.WindowingStrategy;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollection.IsBounded;
+import org.apache.beam.sdk.values.TimestampedValue;
+import org.apache.beam.sdk.values.TimestampedValue.TimestampedValueCoder;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.joda.time.ReadableDuration;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A testing input that generates an unbounded {@link PCollection} of elements, advancing the
+ * watermark and processing time as elements are emitted. After all of the specified elements are
+ * emitted, ceases to produce output.
+ *
+ * <p>Each call to a {@link TestStream.Builder} method will only be reflected in the state of the
+ * {@link Pipeline} after each method before it has completed and no more progress can be made by
+ * the {@link Pipeline}. A {@link PipelineRunner} must ensure that no more progress can be made in
+ * the {@link Pipeline} before advancing the state of the {@link TestStream}.
+ */
+public final class TestStream<T> extends PTransform<PBegin, PCollection<T>> {
+  private final List<Event<T>> events;
+  private final Coder<T> coder;
+
+  /**
+   * Create a new {@link TestStream.Builder} with no elements and watermark equal to {@link
+   * BoundedWindow#TIMESTAMP_MIN_VALUE}.
+   */
+  public static <T> Builder<T> create(Coder<T> coder) {
+    return new Builder<>(coder);
+  }
+
+  private TestStream(Coder<T> coder, List<Event<T>> events) {
+    this.coder = coder;
+    this.events = checkNotNull(events);
+  }
+
+  public Coder<Event<T>> getEventCoder() {
+    return EventCoder.of(coder);
+  }
+
+  /**
+   * An incomplete {@link TestStream}. Elements added to this builder will be produced in sequence
+   * when the pipeline created by the {@link TestStream} is run.
+   */
```

Comment:
Should we make this builder follow the immutable pattern where a new builder instance is returned for each modification, similar to how we build PTransforms like ParDo?